docs/connector-development/connector-builder-ui/record-processing.mdx

```diff
@@ -0,0 +1,339 @@
+```
+{
+    "data": [
+        {
+            "id": "00000000-0000-0000-0000-000000000000",
+            "status": "SENT",
+            "name": "Foo",
+            "subject": "Bar",
+            "from": {
+                "name": "John Doe",
+                "email_address": "john.doe@gmail.com"
+            },
+            "created_at": "2023-04-13T15:28:37+00:00",
+            "sent_at": "2023-04-14T15:28:37+00:00"
+        },
+    ]
+}
+```
+
+To apply a hash function to it, set the "Path" to "`from`, `name`" to select the name property nested in the from object and set the value to `{{ record['from']['name'] | hash('md5') }}`. This hashes the name in the record:
+```
+{
+    "id": "00000000-0000-0000-0000-000000000000",
+    "status": "SENT",
+    "name": "Foo",
+    "subject": "Bar",
+    "from": {
+        "name": "4c2a904bafba06591225113ad17b5cec",
+        "email_address": "john.doe@gmail.com"
+    },
+    "created_at": "2023-04-13T15:28:37+00:00",
+    "sent_at": "2023-04-14T15:28:37+00:00"
+}
+```
+
+Another common use case of the "add" transformation is the enriching of records with their parent resource - check out the [partitioning documentation](/connector-development/connector-builder-ui/partitioning#adding-the-partition-value-to-the-record) for more details.
+
+It's not recommended to use this feature to do projections (like concatenating firstname and lastname into a single "name" field) - in most cases it's beneficial to leave these tasks to a later stage in the data pipeline.
+
+## Meta data
+
+Besides bringing the records in the right shape, it's important to communicate some pieces of meta data about records to the downstream system so they can be handled properly.
+
+### Primary key
+
+The "Primary key" field specifies how to uniquely identify a record. This is important for downstream de-duplication of records (e.g. by the [incremental sync - deduped history sync mode](/understanding-airbyte/connections/incremental-deduped-history)).
+
+In a lot of cases, like for the EmailOctopus example from above, there is a dedicated id field that can be used for this purpose. It's important that the value of the id field is guaranteed to only occur once for a single record.
+
+In some cases there is no such field but a combination of multiple fields is guaranteed to be unique, for example the shipping zone locations of the [Woocommerce API](https://woocommerce.github.io/woocommerce-rest-api-docs/#shipping-zone-locations) do not have an id, but each combination of the `code` and `type` fields is guaranteed to be unique:
+```
+[
+  {
+    "code": "BR",
+    "type": "country"
+  },
+  {
+    "code": "DE",
+    "type": "country"
+  },
+]
+```
+
+In this case, the "Primary key" can be set to "`code`, `type`" to allow automatic downstream deduplication of records based on the value of these two fields.
+
+### Declared schema
```

Comment:
We encountered another edge case during the discourse build a connector session.

If a field is not set on any of the records read, the type will show up as "null", which is almost certainly wrong. Developers should make sure there's at least one valid type for all fields in the schema (and the UI should probably have a warning).

Updating the UI is out of scope here, but we should document this gotcha

Reply:
Good point, added a paragraph for this